### PR TITLE
Refactor KeyboardHelper out of WlKeyboard

### PR DESCRIPTION
**Pull request by wmww**
One of the initially unimplemented components of the input method protocol is keyboard grabs. Squeekboard doesn't grab the keybaord, but CJK input methods (such as [wlanthy](https://github.com/st3r4g/wlanthy) do need to. `zwp_input_method_keyboard_grab_v2` is a separate interface than `wl_keyboard` and will need a separate implementation, but requires much of the same functionality.

After several attempts involving inheritance, I settled on a more composition-oriented approach. `KeybaordHelper` implements the common keyboard logic, and calls into `KeyboardImpl` to send events. In this PR `WlKeyboard` is the only `KeyboardImpl`, but an upcoming PR will add a new grab keybaord implementation.

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
This builds on the changes that appear to be the cause of #2150 - let's not merge until we've addressed that.

**Comment by wmww** [comment]
This is now ready for review again

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
It took me two passes to understand how this fits together. With that understanding it does make sense, but I wonder how we can make it easier to understand.

I've made some suggestions that I hope would have helped me.

Other opinions invited!

File: src/server/frontend_wayland/keyboard_helper.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    virtual void send_keymap_xkb_v1(mir::Fd const& fd, size_t length) = 0;
+    virtual void send_key(uint32_t timestamp, int scancode, bool down) = 0;
+    virtual void send_modifiers(uint32_t depressed, uint32_t latched, uint32_t locked, uint32_t group) = 0;
+
+private:
+    KeyboardImpl(KeyboardImpl const&) = delete;
+    KeyboardImpl& operator=(KeyboardImpl const&) = delete;
+};
+
+class KeyboardHelper
```
[the remark this answers is not in the document]
Comment (wmww) [reply src/server/frontend_wayland/keyboard_helper.h:62]: `KeyboardImpl` too contentious, how about `KeyboardHelper` and `KeyboardCallbacks`?

File: src/server/frontend_wayland/wl_seat.cpp [hunk 2]
```diff
@@ -204,6 +202,11 @@ void mf::WlSeat::for_each_listener(wl_client* client, std::function<void(WlTouch
     touch_listeners->for_each(client, func);
 }
 
+auto mf::WlSeat::make_keyboard_helper(KeyboardImpl* impl) -> std::unique_ptr<KeyboardHelper>
+{
+    return std::make_unique<KeyboardHelper>(impl, keymap, seat, enable_key_repeat);
+}
+
```
[the remark this answers is not in the document]
Comment (wmww) [reply src/server/frontend_wayland/wl_seat.cpp:209]: And also `seat`, but yes. There could be getters on `WlSeat` for all of them, but I thought it cleaner and more maintainable to make `WlSeat` responsible for creating keyboard helpers.

File: src/server/frontend_wayland/keyboard_helper.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright © 2018-2019 Canonical Ltd.
```
[the remark this answers is not in the document]
Comment (wmww) [reply src/server/frontend_wayland/keyboard_helper.cpp]: Fixed

File: src/server/frontend_wayland/keyboard_helper.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: Christopher James Halse Rogers <christopher.halse.rogers@canonical.com>
+ *              William Wold <william.wold@canonical.com>
+ */
+
+#ifndef MIR_FRONTEND_BASIC_WL_KEYBOARD_H
+#define MIR_FRONTEND_BASIC_WL_KEYBOARD_H
```
[the remark this answers is not in the document]
Comment (wmww) [reply src/server/frontend_wayland/keyboard_helper.h]: Fixed

File: src/server/frontend_wayland/keyboard_helper.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (wmww) [reply src/server/frontend_wayland/keyboard_helper.h:62]: Changed name and used private inheritance

**Review by AlanGriffiths — APPROVED** [review]
Still not convinced by `KeyboardHelper`, maybe `KeyboardEventHandler`? But won't block.

bors d+

**Comment by wmww** [comment]
Hmm, `KeyboardEventHandler` seems just as nonspecific as `KeyboardHelper` to me. There might be a better name out there, but I haven't thought of it yet.

bors r+

**Pull request merged**